### PR TITLE
Handle long filenames on Windows (fixes #1295)

### LINKED ISSUE #1295: File / Dir name length and Windows
This is not really a bug in syncthing, but one in windows / ntfs to be aware of and (maybe) to work around:

Pathnames in NTFS are usually restricted to 256 chars (260-4: 'C:\' in the beginning and \NULL at the end; it's complicated, see [1] and [2]):
> "In the Windows API (with some exceptions discussed in the following paragraphs), the maximum length for a path is MAX_PATH, which is defined as 260 characters." [2]

So a perfect valid `long255charfilename.ext` with 255 chars won't be synced, because the creation of the temp file `~syncthing~.long255charfilename.ext.tmp` fails in the sync target (the one receiving the file) [3]

If I read the docs [2] correctly, there are "extended-length path for a maximum total path length of 32,767 characters". So all you should have to do is to prefix your paths with "\\\\?\" if using the proper api calls. But be aware that the conversion of '/' to '\' wont work anymore and relative path using dots neither  (Again, all according to the docs, I did not test this, since I don't have a win machine set up for dev at the moment)

A side effect would then probably be, that those files won't be accessible in windows explorer and certain applications.. But the files are there and applications with proper long file name support should be able to read them. Thats what I would propose as the best way to handle it.

[1] http://stackoverflow.com/questions/265769/maximum-filename-length-in-ntfs-windows-xp-and-windows-vista
[2] https://msdn.microsoft.com/en-us/library/aa365247.aspx
[3] The error message: (running on win8.1 x64)
```
[NLXEG] 23:43:22 INFO: Puller (folder "documents", file "Uni\\Lectures\\INI410 - The Neurobiology of Consciousness - Daniel Kiper, Alex Gamma (Daniel Koch, Heather Berlin) - HS2008\\Lecture Sildes\\Lect 10 - Conclusions to Consciousness from Split-Brain Patients and Agnosias [Dr. H. Berlin].pdf"):
dst create: open s='D:\data\Documents\Uni\Lectures\INI410 - The Neurobiology of Consciousness - Daniel Kiper, Alex Gamma (Daniel Koch, Heather Berlin) - HS2008\Lecture Sildes\~syncthing~.Lect 10 - Conclusions to Consciousness from Split-Brain Patients and Agnosias [Dr. H. Berlin].pdf.tmp': Das System kann den angegebenen Pfad nicht finden.
```

### PR DESCRIPTION
**Pull request by calmh**
There are many approaches to this... This one works on the assumption that it's safest for most of the code if the change is done as early as possible, i.e. at config loading time, lest we forget some critical  place where a conversion might need to be done. And it also assumes that we should not put the `\\?\` in the config or show it to the user in the GUI, as they may be scared and confused by it. If we think most Windows users are fine with `\\?\` being visible then we can just add it at config loading and not do the dance with RawPath/Path...

I've tested that this starts on Windows and seems to hash files while mentioning paths starting with `\\?\` in the debug output... Some more Windows testing would be nice.

### CODE REVIEW

File: internal/config/config.go [hunk 1]
```diff
@@ -94,9 +95,20 @@ func (orig FolderConfiguration) Copy() FolderConfiguration {
 	return c
 }
 
+func (f FolderConfiguration) Path() string {
+	// This is intentionally not a pointer method, becuse things like
+	// cfg.Folders[i].Path() should be valid.
+
+	if runtime.GOOS == "windows" && filepath.IsAbs(f.RawPath) && !strings.HasPrefix(f.RawPath, `\\?\`) {
```
Comment (Zillode) [inline internal/config/config.go]: I think this should be ```!strings.HasPrefix(f.RawPath, `\\`)``` to support network paths
E.g. ```\\?\\\192.168.2.20``` does not work currently

Comment (calmh) [inline internal/config/config.go]: Ah

**Comment by calmh** [comment]
Does this work as far back as XP, or do we need to exclude that somehow?

**Comment by AudriusButkevicius** [comment]
So I don't think we need to compute this every time. We should do it at start up and use that. Also, I think we should expand the ~ and prepend it to that too given we support that.

**Comment by calmh** [comment]
How does that look? This time I haven't even tested it at all on Windows. :)

**Comment by AudriusButkevicius** [comment]
I'm still on the road, but I should be back at a proper keyboard later this evening to test this.

**Comment by AudriusButkevicius** [comment]
So after adding a new folder, all the paths suddenly get unset.
Perhaps we should have a folder.Prepare() and call that in config.Folders() as we are building the map.

Also, perhaps (atleast on windows) we should expand all paths to be absolute, and always apply the magic prefix?

**Comment by calmh** [comment]
That's the easiest and safest... Only downside is that it's user visible and we'll get issues for "Syncthing replaced c:\my documents with \\?\c:\my documents" and so on...

**Comment by Zillode** [comment]
inotify would benefit from \\?\ prefixes, however syncthing-tray or other tray balloon apps would not.

**Comment by calmh** [comment]
We could, optionally, let the path be `\\?\`-prefixes *everywhere* and just hide it in the path text field in the GUI :)

**Comment by AudriusButkevicius** [comment]
or we could go back to your Path() version, which solves this nicely

**Comment by calmh** [comment]
Or that. :/

**Comment by calmh** [comment]
Theoretically that commit should still be somewhere in my reflog...

**Comment by calmh** [comment]
So that then... `Path()` is called fairly seldom so I don't think it's a big deal that it exists... But it may be worth it to move the tilde expansion in there and cache it? Or not cache it. Currently we do the tilde expansion into the config file, so the user says `~/Foo` and it becomes `/Users/jb/Foo` after the mandatory restart....

**Comment by calmh** [comment]
I moved some stuff into `Path()` and attempt to make the path absolute. This may fail; I opted to return the unmodified string rather than pass an annoying error around everywhere... The regular folder checks should catch us if the path does not make sense, I hope...

**Pull request merged**